File: python/2020/05/adv.py
SPLIT_IDX = 7
ROWS_MAX = 127
COLUMNS_MAX = 7


def get_input_lines() -> list[str]:
    with open("input.txt") as f:
        return f.read().splitlines()


def calc_code_score(code: str) -> int:
    rows, cols = code[:SPLIT_IDX], code[SPLIT_IDX:]
    # Rows Calculations
    min, max = 0, ROWS_MAX
    for ch in rows:
        row_mid = min + (max - min) // 2
        if ch == "F":
            max = row_mid
        else:
            min = row_mid + 1

    row = min if rows[-1] == "F" else max

    # Cols Calculations
    min, max = 0, COLUMNS_MAX
    for ch in cols:
        col_mid = min + (max - min) // 2
        if ch == "L":
            max = col_mid
        else:
            min = col_mid + 1

    col = min if cols[-1] == "L" else max

    return row * 8 + col


def part_1():
    input = get_input_lines()
    answer = max((calc_code_score(line) for line in input))

    print(f"Part 1 answer is {answer}")


def test():
    test_input = {
        "FBFBBFFRLR": 357,
        "BFFFBBFRRR": 567,
        "FFFBBBFRRR": 119,
        "BBFFBBFRLL": 820,
    }

    for code, score in test_input.items():
        res_score = calc_code_score(code)
        assert score == res_score, f"Left: {score}, Right: {res_score}"

    print("Tests Passed")


test()
part_1()

File: ruby/2020/05/adv.rb
# frozen_string_literal: true

require '../../aoc_base'

# Year 2020 Day 05
class Aoc05 < AocBase
  # Modelling a seat to provide the id
  class Seat
    def initialize(rows, cols)
      @rows = rows
      @cols = cols
    end

    def self.from_string(line)
      rows = line[..6]
      cols = line[-3..]

      Seat.new(rows, cols)
    end

    def to_s
      "rows: #{@rows}, cols: #{@cols}"
    end

    def calc_id
      row = solve_row
      col = solve_col
      row * 8 + col
    end

    def solve_row
      min = 0
      max = 127
      @rows.chars.each do |char|
        case char
        when 'B'
          min = (min + max) / 2 + 1
        when 'F'
          max = (min + max) / 2
        else
          raise "Rows: Invalid char #{char}"
        end
      end

      min
    end

    def solve_col
      min = 0
      max = 7

      @cols.chars.each do |char|
        case char
        when 'R'
          min = (min + max) / 2 + 1
        when 'L'
          max = (min + max) / 2
        else
          raise "Cols: Invalid char '#{char}'"
        end
      end

      min
    end
  end

  def part_one
    @input
      .split(/\n/)
      .map { |line| Seat.from_string(line).calc_id }
      .max
  end

  def part_two
    all_ids = @input
              .split(/\n/)
              .map { |line| Seat.from_string(line).calc_id }
              .sort

    last_id = all_ids[0]
    all_ids[1..].each do |id|
      return id - 1 if id - last_id != 1

      last_id = id
    end
  end

  def do_tests
    input = { 'FBFBBFFRLR' => 357, 'BFFFBBFRRR' => 567, 'FFFBBBFRRR' => 119, 'BBFFBBFRLL' => 820 }
    input.each do |key, value|
      seat = Seat.from_string(key)
      assert_equal seat.calc_id, value
    end
    puts 'tests pass'
  end
end

if __FILE__ == $PROGRAM_NAME
  aoc = Aoc05.new
  aoc.run
end
